test/infrastructure/docker/exp/internal/controllers/dockermachinepool_controller_phases.go

```diff
@@ -0,0 +1,301 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package controllers implements controller functionality.
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"sort"
+	"strings"
+
+	"github.com/blang/semver/v4"
+	"github.com/pkg/errors"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/kind/pkg/cluster/constants"
+
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	expv1 "sigs.k8s.io/cluster-api/exp/api/v1beta1"
+	infrav1 "sigs.k8s.io/cluster-api/test/infrastructure/docker/api/v1beta1"
+	infraexpv1 "sigs.k8s.io/cluster-api/test/infrastructure/docker/exp/api/v1beta1"
+	"sigs.k8s.io/cluster-api/test/infrastructure/docker/internal/docker"
+	"sigs.k8s.io/cluster-api/test/infrastructure/kind"
+	"sigs.k8s.io/cluster-api/util"
+	"sigs.k8s.io/cluster-api/util/labels/format"
+)
+
+// createNewInstances creates a Docker container for each new replica needed for the MachinePool as well as a DockerMachine to represent it.
+func (r *DockerMachinePoolReconciler) createNewInstances(ctx context.Context, cluster *clusterv1.Cluster, machinePool *expv1.MachinePool, dockerMachinePool *infraexpv1.DockerMachinePool) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	labelFilters := map[string]string{dockerMachinePoolLabel: dockerMachinePool.Name}
+
+	machines, err := docker.ListMachinesByCluster(ctx, cluster, labelFilters)
+	if err != nil {
+		return errors.Wrapf(err, "failed to list all machines in the cluster")
+	}
+
+	matchingMachineCount := len(machinesMatchingInfrastructureSpec(ctx, machines, machinePool, dockerMachinePool))
+	numToCreate := int(*machinePool.Spec.Replicas) - matchingMachineCount
+	for i := 0; i < numToCreate; i++ {
+		log.V(2).Info("Creating a new Docker container for machinePool", "machinePool", machinePool.Name)
+		createdMachine, err := createDockerContainer(ctx, cluster, machinePool, dockerMachinePool)
+		if err != nil {
+			return errors.Wrap(err, "failed to create a new docker machine")
+		}
+
+		log.V(2).Info("Creating a new DockerMachine for Docker container", "container", createdMachine.Name())
+		// Note: we want the DockerMachine to have the same name as the container so that we can easily find it later.
+		if err := r.createDockerMachine(ctx, createdMachine.Name(), cluster, machinePool, dockerMachinePool); err != nil {
+			return errors.Wrap(err, "failed to create a new docker machine")
+		}
+	}
+
+	return nil
+}
+
+// createDockerContainer creates a Docker container to serve as a replica for the MachinePool.
+func createDockerContainer(ctx context.Context, cluster *clusterv1.Cluster, machinePool *expv1.MachinePool, dockerMachinePool *infraexpv1.DockerMachinePool) (*docker.Machine, error) {
+	log := ctrl.LoggerFrom(ctx)
```

Comment:
Is this part of docker machine pool machine core logic?

Reply:
Yes, this was refactored from nodepool.go.